File: leetcode/559.maximum-depth-of-n-ary-tree/go/maxdepth.go
package maximumdepthofnarytree

type Node struct {
	Val      int
	Children []*Node
}

func maxDepth(root *Node) int {
	return getDepth(root, 0)
}

func getDepth(node *Node, depth int) int {
	if node == nil {
		return depth
	}

	depth++
	maxDepth := depth

	for _, n := range node.Children {
		localDepth := getDepth(n, depth)

		if localDepth > maxDepth {
			maxDepth = localDepth
		}
	}

	return maxDepth
}

File: leetcode/559.maximum-depth-of-n-ary-tree/ts/maxdepth.ts
class _Node {
    val: number
    children: _Node[]

    constructor(val?: number, children?: _Node[]) {
        this.val = (val === undefined ? 0 : val)
        this.children = (children === undefined ? [] : children)
    }
}


const maxDepth = (root: _Node | null): number => {
    return getDepth(root, 0)
}

const getDepth = (root: _Node | null, depth: number): number => {
    if (root === null) return depth

    depth++
    let maxDepth = depth

    for (const child of root.children) {
        const localDepth = getDepth(child, depth)
        if (localDepth > maxDepth) maxDepth = localDepth
    }

    return maxDepth
}

export default maxDepth
